source/game_sa/CarAI.cpp

```diff
@@ -55,27 +65,113 @@ void CCarAI::AddAmbulanceOccupants(CVehicle* vehicle) {
+                if (wantedLevel > 2) {
+                    CPed* selectedPed;
+                    if (CGeneral::GetRandomNumberInRange(0.0, 1.0) < 0.25) {
+                        selectedPed = driver;
+                    }
+                    if (CGeneral::GetRandomNumberInRange(0.0, 1.0) < 0.25) {
+                        selectedPed = passenger;
+                    }
+                    selectedPed->GiveDelayedWeapon(WEAPON_SHOTGUN, 1'000);
+                }
+                driver->GetIntelligence()->ClearTasks(true, true);
+                driver->GetTaskManager().SetTask(new CTaskComplexCopInCar{ vehicle, passenger, FindPlayerPed(), true }, TASK_PRIMARY_PRIMARY, true);
+                passenger->GetIntelligence()->ClearTasks(true, true);
+                driver->GetTaskManager().SetTask(new CTaskComplexCopInCar{ vehicle, driver, FindPlayerPed(), true }, TASK_PRIMARY_PRIMARY, false);
+                return;
+            }
+
+            if (arg2 || CGeneral::GetRandomNumberInRange(0, 100) < 50) {
+                vehicle->SetupPassenger(0, PED_TYPE_NONE, false, false);
+            }
+        }
+    }
 }
 
 // 0x41BFA0
 void CCarAI::BackToCruisingIfNoWantedLevel(CVehicle* vehicle) {
-    plugin::Call<0x41BFA0, CVehicle*>(vehicle);
+    if (vehicle->vehicleFlags.bIsLawEnforcer) {
+        CWanted* wanted = FindPlayerWanted(-1);
+        if (!wanted->m_nWantedLevel || wanted->BackOff() || CCullZones::NoPolice()) {
+            CCarCtrl::JoinCarWithRoadSystem(vehicle);
+            uint32 bSirenOrAlarm = vehicle->vehicleFlags.bSirenOrAlarm;
+            vehicle->m_autoPilot.m_nCarMission = MISSION_CRUISE;
+            vehicle->m_autoPilot.m_nCarDrivingStyle = DRIVING_STYLE_STOP_FOR_CARS;
+            vehicle->vehicleFlags.bSirenOrAlarm = bSirenOrAlarm;
+            if (CCullZones::NoPolice()) {
+                vehicle->m_autoPilot.m_nCarMission = MISSION_NONE;
+            }
+        }
+    }
 }
 
 // 0x41C050
 void CCarAI::CarHasReasonToStop(CVehicle* vehicle) {
-    plugin::Call<0x41C050, CVehicle*>(vehicle);
+    vehicle->m_autoPilot.m_nTimeToStartMission = CTimer::m_snTimeInMilliseconds;
 }
 
 // 0x41CD00
-void CCarAI::EntitiesGoHeadOn(CEntity* entity1, CEntity* entity2) {
-    plugin::Call<0x41CD00, CEntity*, CEntity*>(entity1, entity2);
+// ?? it returns something
+bool CCarAI::EntitiesGoHeadOn(CEntity* entity1, CEntity* entity2) {
+    CVector position1 = entity1->m_matrix ? entity1->m_matrix->GetPosition() : entity1->m_placement.m_vPosn;
+    CVector position2 = entity2->m_matrix ? entity2->m_matrix->GetPosition() : entity2->m_placement.m_vPosn;
+    CVector positionDiff = position1 - position2;
+    positionDiff.Normalise();
+
+    CVector forward1 = entity1->m_matrix ? entity1->m_matrix->GetForward() : CVector(-sin(entity1->m_placement.m_fHeading), cos(entity1->m_placement.m_fHeading), 0.0);
+    if (forward1.Dot(positionDiff) > -0.80000001) {
```

Comment:
What is -0.80000001, something special?

Reply:
It's floating point representation of `-0.8`, you can safely round to it. Also **always** use `f` suffix for decimal numbers.
```suggestion
    if (forward1.Dot(positionDiff) > -0.8f) {
```

Reply:
Yeah, floats are inexact, as you might know. So sometimes you get these weird numbers.
As yukani said, just round them to make it look nice.